### PR TITLE
New interface

### LINKED ISSUE #118: Q: how to use the new `use_contract!` macro
Q: I was upgrading parity-ethereum to use the latest ethabi (v5 => v6) and there are a bunch of use_contract! invocations that pass the $name. Without it however, there are many errors like this:

767 |             registrar: registry::Registry::default(),
    |                                  ^^^^^^^^ Could not find `Registry` in `registry`

So I wonder, what is the recommended way to fix this? How should consuming code access the contract?

(the macro was changed here: https://github.com/paritytech/ethabi/pull/107)

### LINKED ISSUE #119: fix 6.0.0 interface
from https://github.com/paritytech/ethabi/pull/85#issuecomment-416182395

> as I was trying to update parity to version `6.0.0` I've noticed that this pull request is fundamentally broken. In particular, `ContractFunction` trait. It's not possible to decode function output without encoding it first! I'm going to revert this part of the pull request

### PR DESCRIPTION
**Pull request by debris**
closes #118 
closes #119 

still todo: https://github.com/paritytech/ethabi/issues/119#issuecomment-416511101

cc @dvdplm @tomusdrw

### CODE REVIEW

File: derive/src/function.rs [hunk 1]
```diff
@@ -0,0 +1,456 @@
...
+								name: #name.into(),
+								inputs: #recreate_inputs,
+								outputs: #recreate_outputs,
+								constant: #constant,
+							}
+						}
+					}
+				}
+
+				pub struct Decoder(Function);
```
Comment (tomusdrw) [inline derive/src/function.rs]: Would be good to add docs for the generated code.

File: derive/src/function.rs [hunk 2]
```diff
@@ -0,0 +1,456 @@
...
+					let f = Function::default();
+					let tokens = vec![#(#tokenize),*];
+					f.function.encode_input(&tokens).expect(INTERNAL_ERR)
+				}
+
+				pub fn decode_output(output: &[u8]) -> ethabi::Result<#outputs_result> {
+					ethabi::FunctionOutputDecoder::decode(&Decoder(Function::default()), output)
+				}
+
+				pub fn call<#(#declarations),*>(#(#definitions),*) -> (ethabi::Bytes, Decoder) {
```
Comment (tomusdrw) [inline derive/src/function.rs]: I didn't see any higher-level tests for `call`, might be good to add some.

File: derive/src/function.rs [hunk 3]
```diff
@@ -0,0 +1,456 @@
...
+					let f = Function::default();
+					let tokens = vec![#(#tokenize),*];
+					(f.function.encode_input(&tokens).expect(INTERNAL_ERR), Decoder(f))
```
Comment (tomusdrw) [inline derive/src/function.rs]: Why duplicate the code here? Maybe just call `encode_input` (or in future delegate to `Encoder` struct)

File: derive/src/function.rs [hunk 4]
```diff
@@ -0,0 +1,456 @@
...
+		let constant = &self.constant;
+		let outputs_result = &self.outputs.result;
+		let outputs_implementation = &self.outputs.implementation;
+
+		quote! {
+			pub mod #module_name {
+				use ethabi;
+				use super::INTERNAL_ERR;
+
+				struct Function {
```
Comment (tomusdrw) [inline derive/src/function.rs]: Why is this wrapper necessary/needed? Why not store `ethabi::Function` directly?

File: derive/src/event.rs [hunk 5]
```diff
@@ -0,0 +1,399 @@
...
+
+					let e = Event::default();
+					e.event.filter(raw).expect(INTERNAL_ERR)
+				}
+
+				pub fn wildcard_filter() -> ethabi::TopicFilter {
+					filter(#(#wildcard_filter_params),*)
+				}
+
+				pub fn parse_log(log: ethabi::RawLog) -> ethabi::Result<super::super::logs::#camel_name> {
```
Comment (tomusdrw) [inline derive/src/event.rs]: Docs missing.

File: derive/src/event.rs [hunk 6]
```diff
@@ -0,0 +1,399 @@
...
+		quote! {
+			pub mod #name {
+				use ethabi;
+				use super::INTERNAL_ERR;
+
+				struct Event {
+					event: ethabi::Event,
+				}
+
+				impl Default for Event {
```
Comment (tomusdrw) [inline derive/src/event.rs]: Again not really convinced we need a wrapper struct for this

File: derive/src/contract.rs [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+		let functions: Vec<_> = self.functions.iter().map(Function::generate).collect();
+		let events: Vec<_> = self.events.iter().map(Event::generate_event).collect();
+		let logs: Vec<_> = self.events.iter().map(Event::generate_log).collect();
+		quote! {
+			use ethabi;
+			const INTERNAL_ERR: &'static str = "`ethabi_derive` internal error";
+
+			#constructor
+
+			pub mod functions {
```
Comment (tomusdrw) [inline derive/src/contract.rs]: docs

File: derive/src/constructor.rs [hunk 8]
```diff
@@ -0,0 +1,165 @@
+use {quote, ethabi};
+
+use super::{
+	input_names, template_param_type, rust_type, get_template_names, to_token, from_template_param,
+	to_ethabi_param_vec,
+};
+
+pub struct Constructor {
```
Comment (dvdplm) [inline derive/src/constructor.rs]: Some docs here would be good.

File: derive/src/event.rs [hunk 9]
```diff
@@ -0,0 +1,399 @@
+use {quote, syn, ethabi};
+use heck::{SnakeCase, CamelCase};
+
+use super::{rust_type, to_syntax_string, from_token, get_template_names, to_token};
+
+pub struct Event {
```
Comment (dvdplm) [inline derive/src/event.rs]: Docs here

File: derive/src/event.rs [hunk 10]
```diff
@@ -0,0 +1,399 @@
...
+			wildcard_filter_params,
+			filter_declarations,
+			filter_definitions,
+			filter_init,
+		}
+	}
+}
+
+impl Event {
+	pub fn generate_log(&self) -> quote::Tokens {
```
Comment (dvdplm) [inline derive/src/event.rs]: Docs.

**Comment by debris** [comment]
> I like how readable the code and tests are. Still not sure how this is used from e.g. registrar in parity-ethereum – are you planning on changing the macro definition or make a follow-up PR to parity-ethereum?

Thanks @dvdplm . Yes, I do!

> Looks good, although the docs for generated code are missing and imho some of the structs are superfluous.

Thanks for the comments. I removed redundant wrappers and updated the docs for ethabi-derive structs. @tomusdrw do you also want to me to add docs for the generated code? I'm not convinced that this is necessary

File: derive/src/constructor.rs [hunk 11]
```diff
@@ -0,0 +1,131 @@
+use {quote, ethabi};
+
+use super::{
+	input_names, template_param_type, rust_type, get_template_names, to_token, from_template_param,
+	to_ethabi_param_vec,
+};
+
+/// Structure used to generate contract's construtor interface.
```
Comment (dvdplm) [inline derive/src/constructor.rs]: typo: "construtor"

**Comment by tomusdrw** [comment]
@debris Yes, I think it would be good to document the generated code as well, otherwise it may trigger warnings if `#![warn(missing_docs)]` is enabled on the crate that uses the generator (and exposes some parts of it).

I think it will also help people understand what the functions are doing when browsing the documentation.

**Pull request merged**